server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -165,6 +198,36 @@ public boolean moveNext() {
         return true;
     }
 
+    private void fetchKeysFromBuffer() {
+        if (nonMatchingKeys == null) {
+            nonMatchingKeys = new ArrayList<>(buffer.size());
+            for (var values : buffer.entrySet()) {
+                if (values.getValue().v2() == false) {
+                    nonMatchingKeys.add(values.getKey());
+                }
+            }
+        }
+    }
+
+    private boolean emitNullValuesPairs() {
+        Integer key = nonMatchingKeys.getFirst();
+        List<Object[]> values = buffer.remove(key).v1();
```

Comment:
I'd also consider to have a pointer to the position we have emitted, instead of removing elements from the list, I'm guessing it'd be more efficient. This way it would be easier to use a `BitSet` for `nonMatchingKeys` instead of an arraylist.

Reply:
I used an iterator with the values and removed the reinserting.